### PR TITLE
[Feature] Spvcnn backbone

### PR DESCRIPTION
**Pull request by sunjiahao1999**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add spvcnn backbone.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by Xiangxu-0103 — COMMENTED** [review]
Please clean this PR to only support SPVCNN backbone.

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            Defaults to 4.
+        base_channels (int): The input channels for first encoder layer.
+            Defaults to 32.
+        encoder_channels (List[int]): Convolutional channels of each encode
+            layer. Defaults to [32, 64, 128, 256].
+        decoder_channels (List[int]): Convolutional channels of each decode
+            layer. Defaults to [256, 128, 96, 96].
+        num_stages (int): Number of stages in encoder and decoder.
+            Defaults to 4.
+        drop_ratio: Dropout ratio of voxel features. Defaults to 0.3.
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/spvcnn_backone.py:38]:
```suggestion
        drop_ratio (float): Dropout ratio of voxel features. Defaults to 0.3.
```

File: mmdet3d/models/data_preprocessors/data_preprocessor.py [hunk 2]
```diff
@@ -429,19 +456,57 @@ def get_voxel_seg(self, res_coors: torch.Tensor, data_sample: SampleList):
...
             data_sample: (:obj:`Det3DDataSample`): The annotation data of
-                every samples. Add voxel-wise annotation forsegmentation.
+                every samples. Add voxel-wise annotation for segmentation.
         """
+        pts_semantic_mask = data_sample.gt_pts_seg.pts_semantic_mask
+        voxel_semantic_mask, _ = dynamic_scatter_3d(
+            F.one_hot(pts_semantic_mask.long()).float(),
+            res_coors.contiguous(), 'mean')
+        voxel_semantic_mask = torch.argmax(voxel_semantic_mask, dim=-1)
+        data_sample.gt_pts_seg.voxel_semantic_mask = voxel_semantic_mask
```
Comment (Xiangxu-0103) [inline mmdet3d/models/data_preprocessors/data_preprocessor.py:466]: Whether this modification will cause bc-breaking for cylinder3D?

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        inserted_feat = F.spvoxelize(z.F, idx_query, counts)
+
+        new_tensor = SparseTensor(inserted_feat, inserted_coords, 1)
+        new_tensor.cmaps.setdefault(new_tensor.stride, new_tensor.coords)
+        z.additional_features['idx_query'][1] = idx_query
+        z.additional_features['counts'][1] = counts
+        return new_tensor
+
+    def voxel_to_point(self,
+                       x: SparseTensor,
```
Comment (JingweiZhang12) [inline mmdet3d/models/backbones/spvcnn_backone.py:139]: x > voxels,  z > points

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            new_tensor = PointTensor(
+                new_feat, z.C, idx_query=z.idx_query, weights=z.weights)
+            new_tensor.additional_features = z.additional_features
+
+        if point_transform is not None:
+            new_tensor.F = new_tensor.F + point_transform(z.F)
+
+        return new_tensor
+
+    def point_to_voxel(self, x: SparseTensor, z: PointTensor) -> SparseTensor:
```
Comment (JingweiZhang12) [inline mmdet3d/models/backbones/spvcnn_backone.py:192]: The same as the above

**Review by JingweiZhang12 — APPROVED** [review]

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+        Args:
+            voxel_features (Tensor): Voxel features in shape (N, C).
+            coors (Tensor): Coordinates in shape (N, 4),
+                the columns in the order of (x_idx, y_idx, z_idx, batch_idx).
+
+        Returns:
+            PointTensor: Backbone features.
+        """
+        x = SparseTensor(voxel_features, coors)
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/spvcnn_backone.py:83]: Can not guess what `x` and `z` mean, seems that we should rename it more clear, such as `voxels` and `points` or others.

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            nearest (bool): Whether to use nearest neighbor interpolation.
+                Defaults to False.
+
+        Returns:
+            PointTensor: Points with new features.
+        """
+        if points.idx_query is None or points.weights is None or \
+                points.idx_query.get(voxels.s) is None or \
+                points.weights.get(voxels.s) is None:
+            off = get_kernel_offsets(2, voxels.s, 1, device=points.F.device)
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/spvcnn_backone.py:159]: off -> offset?

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            pc_hash = F.sphash(voxels.C.to(points.F.device))
+            idx_query = F.sphashquery(old_hash, pc_hash)
+            weights = F.calc_ti_weights(
+                points.C, idx_query,
+                scale=voxels.s[0]).transpose(0, 1).contiguous()
+            idx_query = idx_query.transpose(0, 1).contiguous()
+            if nearest:
+                weights[:, 1:] = 0.
+                idx_query[:, 1:] = -1
+            new_feat = F.spdevoxelize(voxels.F, idx_query, weights)
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/spvcnn_backone.py:174]: new_feat -> new_features

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+                points.C,
+                idx_query=points.idx_query,
+                weights=points.weights)
+            new_tensor.additional_features = points.additional_features
+            new_tensor.idx_query[voxels.s] = idx_query
+            new_tensor.weights[voxels.s] = weights
+            points.idx_query[voxels.s] = idx_query
+            points.weights[voxels.s] = weights
+        else:
+            new_feat = F.spdevoxelize(voxels.F, points.idx_query.get(voxels.s),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/spvcnn_backone.py:186]: new_feat -> new_features

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+            sparse_hash = F.sphash(voxels.C)
+            idx_query = F.sphashquery(pc_hash, sparse_hash)
+            counts = F.spcount(idx_query.int(), voxels.C.shape[0])
+            points.additional_features['idx_query'][voxels.s] = idx_query
+            points.additional_features['counts'][voxels.s] = counts
+        else:
+            idx_query = points.additional_features['idx_query'][voxels.s]
+            counts = points.additional_features['counts'][voxels.s]
+
+        inserted_feat = F.spvoxelize(points.F, idx_query, counts)
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/spvcnn_backone.py:228]: inserted_feat -> inserted_features

File: mmdet3d/models/backbones/spvcnn_backone.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+            decoder_outs.append(x)
+            if i == 1:
+                z = self.voxel_to_point(x, z, self.point_transforms[1])
+                x = self.point_to_voxel(x, z)
+                x.F = self.dropout(x.F)
+
+        z = self.voxel_to_point(x, z, self.point_transforms[2])
+        return z
+
+    def initial_voxelize(self, z: PointTensor) -> SparseTensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/spvcnn_backone.py:114]: z -> points or other more clear name.

**Review by Xiangxu-0103 — APPROVED** [review]

**Pull request merged**